qt: Cleanup MacDockIconHandler class

hebasto opened the pull request:
The Objective-C `MacDockIconHandler` class currently does such jobs on macOS:

1. Set the Dock icon
2. Support the Dock icon menu
3. Handle the Dock icon click event

Qt does the first job natively.
Qt has native support the Dock icon menu only since version 5.2
Qt does not handle `applicationShouldHandleReopen` event. And this brakes all Qt support for the Dock icon click event.

This PR:
- removes Objective-C code for the Dock icon setting. Qt `setWindowIcon()` does this work
- removes Objective-C code for the Dock icon menu. Qt `setAsDockMenu()` does this work
- uses Qt signal for the Dock icon click event

promag commented:
Concept ACK.

practicalswift commented:
Concept ACK

Net change: -189 lines. Nice cleanup!

fanquake commented:
Concept ACK

There's at least one bug here. Using e0d1883, if you close the main window using the red cross, you can no longer bring the window back by clicking on the dock icon. 
This might be solved by one of the other GUI related PRs, possibly #14123.

ismail120572 commented:
Who do that

Ismail

Pada tanggal Sel, 30 Okt 2018 18:02, fanquake <notifications@github.com>
menulis:

> Concept ACK
>
> There's at least one bug here. Using e0d1883
> <https://github.com/bitcoin/bitcoin/commit/e0d1883d0e98c265db09b6a48b9c124f45edf898>,
> if you close the main window using the red cross, you can no longer bring
> the window back by clicking on the dock icon.
> This might be solved by one of the other GUI related PRs, possibly #14123
> <https://github.com/bitcoin/bitcoin/pull/14123>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-434257329>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ApWp_A5jtWwHnwL2Ot_VkmHTBNCHlDqMks5uqDHVgaJpZM4X-Owd>
> .
>

ismail120572 commented:
Ismail

Pada tanggal Sel, 30 Okt 2018 18:02, fanquake <notifications@github.com>
menulis:

> Concept ACK
>
> There's at least one bug here. Using e0d1883
> <https://github.com/bitcoin/bitcoin/commit/e0d1883d0e98c265db09b6a48b9c124f45edf898>,
> if you close the main window using the red cross, you can no longer bring
> the window back by clicking on the dock icon.
> This might be solved by one of the other GUI related PRs, possibly #14123
> <https://github.com/bitcoin/bitcoin/pull/14123>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-434257329>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ApWp_A5jtWwHnwL2Ot_VkmHTBNCHlDqMks5uqDHVgaJpZM4X-Owd>
> .
>

promag commented:
> This might be solved by one of the other GUI related PRs, possibly #14123.

@fanquake no it doesn't solve it. But using a dock menu action, like "Send", it reopens.

hebasto committed:
Remove obj_c for macOS Dock icon setting

Qt `setWindowIcon()` does this work.

hebasto committed:
Use Qt signal for macOS Dock icon click event

This moves the Dock icon click reaction code to the common place and
allows some cleanup in obj_c code.

According to the Apple's docs `class_replaceMethod` behaves as
`class_addMethod`, if the method identified by name does not yet exist;
or as `method_setImplementation`, if it does exist.

hebasto commented:
@fanquake @promag @practicalswift 
Thank you for your reviews.
The original goal of this PR (to completely remove `MacDockIconHandler` class) has revealed inaccessible.
So PR has be reworked. Would you mind re-reviewing?

fanquake commented on src/qt/macdockiconhandler.mm:
```diff
@@ -49,61 +44,22 @@ void setupDockClickHandler() {
     this->m_dummyWidget = new QWidget();
     this->m_dockMenu = new QMenu(this->m_dummyWidget);
-    this->setMainWindow(nullptr);
 #if QT_VERSION >= 0x050200
```
You can drop the `QT_VERSION` check here

hebasto committed:
Remove obj_c for macOS Dock icon menu

Qt `setAsDockMenu()` does this work.

hebasto commented:
@fanquake Your comment is addressed. Please re-review.

Sjors commented on src/qt/macdockiconhandler.mm:
```diff
@@ -34,74 +28,13 @@ void setupDockClickHandler() {
-            class_replaceMethod(delClass, shouldHandle, (IMP)dockClickHandler, "B@:");
-        else
-            class_addMethod(delClass, shouldHandle, (IMP)dockClickHandler,"B@:");
+        class_replaceMethod(delClass, shouldHandle, (IMP)dockClickHandler, "B@:");
```
Why did/do we need this `class_replaceMethod` stuff?

hebasto replied:
This implements the method of the application delegate which handles the `reopen application` event (i.e., the user clicks application icon in the Dock).
Ref: [How Cocoa Applications Handle Apple Events](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ScriptableCocoaApplications/SApps_handle_AEs/SAppsHandleAEs.html)

Sjors reviewed (approved):
tACK efb681d on macOS 10.14.1, thanks for reducing the amount of OS specific code we need to reason about!

hebasto commented:
Fixed bug with hided/inactive/obscured main window: `toggleHidden()` does not work on macOS properly.

jonasschnelli commented:
This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9.

hebasto commented:
@jonasschnelli 
> This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9.


https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes.md:
> From 0.17.0 onwards macOS <10.10 is no longer supported. 0.17.0 is built using Qt 5.9.x, which doesn't support versions of macOS older than 10.10.

What is the reason to choose so wide range of the tested platforms?

ismail120572 commented:
Dikirim dari ponsel cerdas Samsung Galaxy saya.
-------- Pesan asli --------Dari: Jonas Schnelli <notifications@github.com> Tanggal: 04/11/18  21:03  (GMT+07:00) Ke: bitcoin/bitcoin <bitcoin@noreply.github.com> Cc: Ismail <nusagala@gmail.com>, Comment <comment@noreply.github.com> Subjek: Re: [bitcoin/bitcoin] qt: Cleanup MacDockIconHandler class (#14597) 
This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/bitcoin/bitcoin","title":"bitcoin/bitcoin","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/bitcoin/bitcoin"}},"updates":{"snippets":[{"icon":"PERSON","message":"@jonasschnelli in #14597: This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9."}],"action":{"name":"View Pull Request","url":"https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-435672243"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-435672243",
"url": "https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-435672243",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
},
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "AF6C5A86-E920-430C-9C59-A73278B5EFEB",
"title": "Re: [bitcoin/bitcoin] qt: Cleanup MacDockIconHandler class (#14597)",
"sections": [
{
"text": "",
"activityTitle": "**Jonas Schnelli**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@jonasschnelli",
"facts": [

]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"bitcoin/bitcoin\",\n\"issueId\": 14597,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close pull request",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"PullRequestClose\",\n\"repositoryFullName\": \"bitcoin/bitcoin\",\n\"pullRequestId\": 14597\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-435672243"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 402189341\n}"
}
],
"themeColor": "26292E"
}
]

ismail120572 commented:
tidak sma sekali tidak keberatan justru saya senang ada yang meperhatian
seperti anda hingga terbuka transapasi
sekali lagi terima kasisih

Pada tanggal Kam, 1 Nov 2018 pukul 04.33 Hennadii Stepanov <
notifications@github.com> menulis:

> @fanquake <https://github.com/fanquake> @promag
> <https://github.com/promag> @practicalswift
> <https://github.com/practicalswift>
> Thank you for your reviews.
> The original goal of this PR (to completely remove MacDockIconHandler
> class) has revealed inaccessible.
> So PR has be reworked. Would you mind re-reviewing?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-434836662>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ApWp_MOEqRZDBFVBeyfNRzZasH9eKzHEks5uqgkYgaJpZM4X-Owd>
> .
>

jonasschnelli commented:
> > This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9.
> https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes.md:

The "official" binaries are using Qt5.9 which doesn't mean that we drop support self-compilation with an older Qt version.

We should at least do some tests on something between OSX 10.10 and 10.14 with Qt5.5 and Qt5.9 (and not just only 10.14&Qt5.9).
